### PR TITLE
Terraform 1

### PR DESCRIPTION
**Pull request by gitkvark**
## Initial commit for terraform deployment

This setup will install the preprod environment:
- creation of the EKS cluster
- setup of the loadbalancing bits (ELB and Traefik)
- setup of the Helm deployments (cert-manager, metrics-server, traefik and our fastapi-app)

Setup needed to execute the cluster creation:

### Terraform CLI:
- set up the AWS credential environmental variables (AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY)
- check out the code
- CHDIR to the terraform/provisitioning/preprod directory
- execute "terraform init" and "terraform apply"
- once complete, CHDIR to terraform/deployments/preprod directory
- execute "terraform init" and "terraform apply"

To change the name of the cluster ("devops-boot-preprod" by default), use the argument -var=cluster_name=your_cluster_name

```terraform apply -var=cluster_name=your_cluster_name```

### Terraform Cloud:
- create the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY secrets on the UI.
- either install the Terraform Cloud app in Github or set up Github Actions to call Terraform Cloud (TODO: details to be added)
- review the plan and apply it (re-run it once more manually due to the issue described below)

### Post-install steps:
Terraform creates a new ELB with a new domain name. In order for the devops-fastapi-staging.kvark.fr domain to function correctly, it needs to point to the ELB endpoint. The endpoint can be listed using this command, once the cluster is operational:

First let's get the cluster's config:
```
$ aws eks update-kubeconfig --region eu-west-3 --name devops-boot-preprod --kubeconfig ~/.kube/devops-boot-preprod.config
$ export KUBECONFIG=~/.kube/devops-boot-preprod.config
```
Then let's get the endpoint:
```
$ kubectl get service -n traefik
NAME      TYPE           CLUSTER-IP      EXTERNAL-IP                                                                     PORT(S)                      AGE
traefik   LoadBalancer   172.20.49.101   acc642e764d38481fa0ff58c44d01605-39d2472a0af14d46.elb.eu-west-3.amazonaws.com   80:30457/TCP,443:31983/TCP   92m
```
The domain needs to be a CNAME to the domain name listed above.

### CODE REVIEW

File: terraform/modules/eks/main.tf [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Copyright (c) HashiCorp, Inc.
+# SPDX-License-Identifier: MPL-2.0
+
```
Comment (silopolis) [inline terraform/modules/eks/main.tf:3]: can be removed I believe

**Review by silopolis — COMMENTED** [review]
Looks very nice to me!
Now I need to run it...

Keep pushing fixes & improvements but keep the scope as is 'cause it's already something to swallow ;)

Could you add a bare docs/installation/README to it so we know where to where/how to start this for testing.
I'll convert it into proper docs once the PR tested and merged

File: terraform/modules/eks/main.tf (hunk 1 above)
Comment (gitkvark) [reply terraform/modules/eks/main.tf:3]: Oh yes.

**Review by silopolis — APPROVED** [review]
I wish there have been a bit more comments, but let's move forward...

**Review by JoAngel8 — APPROVED** [review]
review done, test done, perfect.

**Pull request merged**